### PR TITLE
Use protos for Invocations

### LINKED ISSUE #1249: Replace `Invocation` struct with protobuf
https://github.com/project-oak/oak/blob/7b89fbdada697230672eae48b10fbe7904a33b2f/sdk/rust/oak/src/grpc/invocation.rs#L17-L22

Now that we have support for handles in protobuf, we should replace data structures that we currently manually serialize, with alternative protobuf versions.

@conradgrobler would you like to look into it, since we talked about this before?

### PR DESCRIPTION
**Pull request by conradgrobler**
This PR replaces Invocation structs in the SDK and Runtime with protos. It also enables using protobuf handle serialisation in the runtime.

Sumary of changes:

- Replace Invocation structs with protos
- Add GrpcInvocationSender and HttpInvocationSender protos for sending the Invocation channel during startup
- Update c++ modules to support GrpcInvocationSender
- Use updated versions of externally stored binaries for aggregator and trusted information retrieval examples

Fixes #1249

Unblocks #1246 

# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: oak_runtime/src/io.rs [hunk 1]
```diff
@@ -22,13 +22,23 @@ use oak_abi::{ChannelReadStatus, OakStatus};
 use oak_io::{Decodable, Encodable, OakError};
 pub use oak_io::{Receiver, Sender};
 
+/// Wrapper for storing a decodable message and the size of the original encoded version.
+pub struct MessageWithSize<T> {
+    pub message: T,
+    pub size: usize,
+}
+
```
Comment (tiziano88) [inline oak_runtime/src/io.rs:30]: Why is this necessary? And what happens if the value of the size field does not match the actual size? AFAICT this is only used for metrics, and I feel the extra complexity is not warranted. We can look into the message and see its size there, if needed?

File: oak_runtime/src/node/grpc/invocation.rs [hunk 2]
```diff
@@ -16,70 +16,54 @@
...
-            }),
-            sender: Sender::new(WriteHandle {
-                handle: message.handles[1],
-            }),
-        })
-    }
-}
-
 impl Invocation {
     pub fn close(self, runtime: &RuntimeProxy) {
```
Comment (tiziano88) [inline oak_runtime/src/node/grpc/invocation.rs:30]: Could you add a comment saying that this ignores any errors while closing the underlying channels?

File: oak_runtime/src/node/grpc/invocation.rs [hunk 3]
```diff
@@ -16,70 +16,54 @@
...
+            }
+            None => error!("No sender on invocation."),
+        };
     }
+
     pub fn receive_request(&self, runtime: &RuntimeProxy) -> Result<GrpcRequest, OakError> {
-        self.receiver.receive(runtime)
+        match &self.receiver {
+            Some(receiver) => receiver.receive(runtime),
+            None => Err(OakError::OakStatus(OakStatus::ErrBadHandle)),
```
Comment (tiziano88) [inline oak_runtime/src/node/grpc/invocation.rs:52]: I thought this should be a similar error to above when no receiver or sender is present. Not exactly the same, but they should be related?

File: oak_runtime/src/node/grpc/server/mod.rs [hunk 4]
```diff
@@ -574,7 +537,7 @@ impl Drop for MetricsRecorder {
 
 struct GrpcResponseIterator {
     runtime: RuntimeProxy,
-    response_reader: oak_abi::Handle,
+    response_reader: ReadHandle,
```
Comment (tiziano88) [inline oak_runtime/src/node/grpc/server/mod.rs:540]: Should this field be of type `Receiver`? Or should it be renamed to `_handle`?

File: oak_runtime/src/io.rs (hunk 1 above)
Comment (conradgrobler) [reply oak_runtime/src/io.rs:30]: This was to be able to use a Receiver rather than manually wait on channels, receive and decode the message to simplify the logic in the GrpcResponseIterator. I assumed there might be a need in future other cases might als use the size of the messages for monitoring, so thought it would be useful to make it possible to use a Receiver in these cases.

I am not sure what you mean by looking into the message and see its size, but I assume you mean go back to manually waiting on the channels and reading the raw message. Is that correct?

File: oak_runtime/src/node/grpc/invocation.rs (hunk 3 above)
Comment (conradgrobler) [reply oak_runtime/src/node/grpc/invocation.rs:52]: The above function logs the error as it does not return a result, where this returns a result, so I thought it would be sufficient to propagate the error to be logged at a higher level. Do you suggest logging here and also returning the error or something else?

File: oak_runtime/src/node/grpc/server/mod.rs (hunk 4 above)
Comment (conradgrobler) [reply oak_runtime/src/node/grpc/server/mod.rs:540]: If I move back to waiting on channels and reading raw messages I will change this back to being a raw Handle and rename the field.

File: sdk/rust/oak/build.rs [hunk 5]
```diff
@@ -17,12 +17,13 @@
...
     oak_utils::compile_protos_with_options(
         &[
+            "../../../oak_services/proto/grpc_invocation.proto",
+            "../../../oak_services/proto/http_invocation.proto",
             "../../../oak_services/proto/storage_service.proto",
             "../../../oak_services/proto/roughtime_service.proto",
         ],
         &["../../.."],
         oak_utils::ProtoOptions {
-            derive_handle_visit: false,
```
Comment (ipetr0v) [inline sdk/rust/oak/build.rs:25]: Is `handle_visit` not necessary here? What does it do?

File: sdk/rust/oak/src/grpc/mod.rs [hunk 6]
```diff
@@ -243,8 +251,8 @@ where
     // Build an Invocation holding the two channels and send it down the
     // specified channel.
     let invocation = Invocation {
-        request_receiver: req_receiver.clone(),
-        response_sender: rsp_sender.clone(),
+        receiver: Some(req_receiver.clone()),
```
Comment (ipetr0v) [inline sdk/rust/oak/src/grpc/mod.rs:254]: Does `GrpcInvocationSender` check that both fields are not `None`? Since it will be an internal error

**Review by ipetr0v — APPROVED** [review]

File: oak_runtime/src/node/grpc/invocation.rs [hunk 7]
```diff
@@ -16,70 +16,54 @@
...
-                handle: message.handles[0],
-            }),
-            sender: Sender::new(WriteHandle {
-                handle: message.handles[1],
-            }),
-        })
-    }
-}
-
 impl Invocation {
```
Comment (rbehjati) [inline oak_runtime/src/node/grpc/invocation.rs:29]: Could this and a similar implementation of HttpInvocation in `sdk/rust/oak/src/http/mod.rs` be unified into a generic implementation?

File: oak_io/src/handle.rs [hunk 8]
```diff
@@ -187,12 +187,15 @@ macro_rules! handle_visit_blanket_impl {
 // See: https://github.com/danburkert/prost#scalar-values
 handle_visit_blanket_impl!((), f64, f32, i32, i64, u32, u64, bool, String, Vec<u8>);
 
-// Provide an implementation for the Any type in Prost.
-handle_visit_blanket_impl!(prost_types::Any);
+// Provide an implementation for the Any and Timestamp types in Prost.
+handle_visit_blanket_impl!(prost_types::Any, prost_types::Timestamp);
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply oak_io/src/handle.rs:191]: Any types that want to implement Encodable or Decodable must implement HandleVisit. Any objects that will be sent using SenderExt must implement Encodable and any objects received using ReceiverExt must implement Decodable. To be able to send protos around they must all implement HandleVisit.

HandleVisit provides the ability to recursively visit all properties an a struct. This means that all properties that are on a struct that implements HandleVisit must in turn also implement HandleVisit. prost_types::Timestamp is used by the IntrospecitionEvents proto.

In cases where there are no handles there is nothing to do, but the trait must still be implemented. So the handle_visit_blanket_impl macro just implements a no-op to satisfy the type constraint.

File: sdk/rust/oak/build.rs (hunk 5 above)
Comment (conradgrobler) [reply sdk/rust/oak/build.rs:25]: By removing this line the HandleVisit trait will be automatically derived for the protos during compilation (the default is true). This is required for the new protos to be sent and received using Sender and Receiver.

File: oak_io/src/handle.rs [hunk 9]
```diff
@@ -187,12 +187,15 @@ macro_rules! handle_visit_blanket_impl {
...
 
 // Provide an implementation for oak_abi type that an implementation cannot be derived for.
 // It does not contain handles, so a blanket impl is sufficient.
-handle_visit_blanket_impl!(oak_abi::proto::oak::application::ConfigMap);
+handle_visit_blanket_impl!(
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply oak_io/src/handle.rs:195]: As noted in another comment in this reply this is a no-op to satisfy the type constraints, as the Label proto is used by the storage service protos.

File: oak_runtime/src/io.rs (hunk 1 above)
Comment (conradgrobler) [reply oak_runtime/src/io.rs:30]: As discused separately, removed this and used the size of the rsp_msg field rather thatn the size of the entire GrpcResponse proto for monitoring.

File: oak_runtime/src/node/grpc/invocation.rs (hunk 2 above)
Comment (conradgrobler) [reply oak_runtime/src/node/grpc/invocation.rs:30]: Added a comment.

@ipetr0v the ignoring of errors was a pre-existing pattern in the function. I assume the reasoning is that the primary reason for the errors would be that the channel was already closed previously, so logging the error should be sufficient.

File: oak_io/src/handle.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (rbehjati) [reply oak_io/src/handle.rs:191]: Thanks for the explanation.

File: sdk/rust/oak/src/grpc/mod.rs (hunk 6 above)
Comment (conradgrobler) [reply sdk/rust/oak/src/grpc/mod.rs:254]: There is not a `send` function on GrpcInvocationSender. Its WriteHandle is extracted and used in `get_invocation_channel` and an Error is returned there if the `sender` is `None`.

File: oak_runtime/src/node/grpc/server/mod.rs (hunk 4 above)
Comment (conradgrobler) [reply oak_runtime/src/node/grpc/server/mod.rs:540]: Changed it to an `Option<Receiver>` (it must be an `Option` to support consuming `receiver` when calling `close` in `drop`)

File: oak_runtime/src/node/grpc/invocation.rs (hunk 7 above)
Comment (conradgrobler) [reply oak_runtime/src/node/grpc/invocation.rs:29]: I don't think this can be done easily. This `GrpcInvocation` implementation is in the Runtime and the implementation of the `HttpInvocation` is in the SDK. The Runtime and SDK use different versions of SenderExt and ReceiverExt with different signatures for `send`, `receive` and `close` to support different mechanisms for interacting with handles depending on whether the code runs in a WASM node or not.

Comment (rbehjati) [reply oak_runtime/src/node/grpc/invocation.rs:29]: hmm! I did not realize that we have two different versions of `SenderExt` and `ReceiverExt `.

**Review by rbehjati — APPROVED** [review]
